File: scripts/python/03_sst_reef_heat.py
"""
03_sst_reef_heat.py  —  SEA SURFACE TEMPERATURE & REEF HEAT (Python / geemap)
SPREP / UNEP GEE Climate Workshop 2026

Python equivalent of javascript/03_sst_reef_heat.js.
Maps recent SST around your islands and plots the monthly SST anomaly
time series (the coral-bleaching heat-stress signal).

Dataset: NOAA CDR OISST v2.1 (NOAA/CDR/OISST/V2_1)
  band 'sst'  scale 0.01 -> °C
  band 'anom' scale 0.01 -> °C anomaly vs 1971-2000
"""

import ee
import geemap
import pandas as pd
import matplotlib.pyplot as plt
# Shared country selector (the 14 SPREP member countries). In Colab / Jupyter
# this helper file may not be present, so fetch it from the workshop repo if so.
try:
    from _pacific_aoi import get_country
except ModuleNotFoundError:
    import urllib.request
    urllib.request.urlretrieve(
        'https://raw.githubusercontent.com/tavitas/sprep-gee-climate-workshop/'
        'main/scripts/python/_pacific_aoi.py', '_pacific_aoi.py')
    from _pacific_aoi import get_country

ee.Initialize(project='your-project-id')   # <-- your registered project

# ===== 1. SETTINGS =====
COUNTRY = 'Fiji'        # any Pacific nation, e.g. 'Tuvalu', 'Marshall Islands'

land = get_country(COUNTRY)                       # ee.Geometry
ocean = land.bounds().buffer(150000)              # ocean AOI ~150 km offshore

oisst = ee.ImageCollection('NOAA/CDR/OISST/V2_1')

# ===== 2. RECENT MEAN SST MAP =====
mean_sst = (oisst.select('sst')
            .filter(ee.Filter.date('2024-01-01', '2025-01-01'))
            .mean().multiply(0.01).clip(ocean))

sst_palette = ['000080', '0000ff', '00ffff', 'ffff00', 'ff8000', 'ff0000', '800000']
vis = {'min': 24, 'max': 31, 'palette': sst_palette}

Map = geemap.Map()
Map.centerObject(ocean, 7)
Map.addLayer(mean_sst, vis, 'Mean SST 2024 (°C)')
Map.add_colorbar(vis, label='Sea surface temperature (°C)')
Map

# ===== 3. MONTHLY SST ANOMALY TIME SERIES =====
start = ee.Date('2015-01-01')
n_months = ee.Date(pd.Timestamp.today().strftime('%Y-%m-01')).difference(start, 'month').floor()

def monthly_anom(m):
    t0 = start.advance(m, 'month')
    t1 = t0.advance(1, 'month')
    img = oisst.select('anom').filter(ee.Filter.date(t0, t1)).mean().multiply(0.01)
    val = img.reduceRegion(ee.Reducer.mean(), ocean, 25000).get('anom')
    return ee.Feature(None, {'date': t0.format('YYYY-MM-dd'), 'anom': val})

fc = ee.FeatureCollection(ee.List.sequence(0, n_months).map(monthly_anom))
data = fc.getInfo()['features']
df = pd.DataFrame([(f['properties']['date'], f['properties']['anom']) for f in data],
                  columns=['date', 'anom'])
df['date'] = pd.to_datetime(df['date'])
df = df.dropna().sort_values('date')

plt.figure(figsize=(11, 4))
plt.plot(df['date'], df['anom'], color='#b2182b', lw=1.5)
plt.axhline(0, color='grey')
plt.axhline(1, color='orange', ls='--', label='+1°C (bleaching watch)')
plt.title(f'Monthly SST anomaly around {COUNTRY} (2015-present)')
plt.ylabel('SST anomaly (°C vs normal)')
plt.legend()
plt.tight_layout()
plt.savefig(f'{COUNTRY}_sst_anomaly.png', dpi=150)
print(f'Saved chart: {COUNTRY}_sst_anomaly.png')
print('Sustained anomalies above +1°C indicate coral bleaching heat stress.')

File: scripts/javascript/03_sst_reef_heat.js
/**********************************************************************
 * 03_sst_reef_heat.js   —  OCEAN, SEA SURFACE TEMPERATURE & REEF HEAT
 * SPREP / UNEP GEE Climate Workshop 2026
 *
 * GOAL: Map sea surface temperature (SST) around your islands and chart
 *       SST anomalies — the warm spikes that drive coral bleaching.
 *
 * DATASET: NOAA CDR OISST v2.1  (NOAA/CDR/OISST/V2_1)
 *   - Daily, 0.25°, global ocean, 1981-present.
 *   - Band 'sst'  = sea surface temp, scale 0.01 -> °C
 *   - Band 'anom' = SST anomaly vs climatology, scale 0.01 -> °C
 *
 * This uses an OCEAN area (a buffer around the country), so it works the
 * same for high islands and atolls.
 **********************************************************************/

// ===== 1. COUNTRY SELECTOR -> OCEAN AREA ============================
var COUNTRY = 'Fiji';
var LSIB_NAMES = {
  'Fiji':'Fiji', 'Papua New Guinea':'Papua New Guinea', 'Solomon Islands':'Solomon Is',
  'Vanuatu':'Vanuatu', 'Samoa':'Samoa'};
var POINT_AOI = {
  'Tonga':[-174.8,-20.0,300000], 'Palau':[134.58,7.5,120000], 'Tuvalu':[178.5,-7.8,350000],
  'Kiribati':[173.0,1.4,500000], 'Nauru':[166.93,-0.52,40000], 'Niue':[-169.87,-19.05,40000],
  'Cook Islands':[-159.78,-21.23,300000], 'Marshall Islands':[169.0,8.0,600000],
  'Federated States of Micronesia':[158.21,6.92,500000]};
var land;
if (LSIB_NAMES[COUNTRY]) {
  land = ee.FeatureCollection('USDOS/LSIB_SIMPLE/2017')
           .filter(ee.Filter.eq('country_na', LSIB_NAMES[COUNTRY])).geometry();
} else {
  var pt = POINT_AOI[COUNTRY];
  land = ee.Geometry.Point([pt[0], pt[1]]).buffer(pt[2]);
}
// Ocean area of interest = land bounding box grown by ~150 km.
var ocean = land.bounds().buffer(150000);
Map.centerObject(ocean, 7);

var sstPalette  = ['000080','0000ff','00ffff','ffff00','ff8000','ff0000','800000'];
var anomPalette = ['2166ac','67a9cf','d1e5f0','f7f7f7','fddbc7','ef8a62','b2182b'];

// ===== 2. RECENT MEAN SST =========================================
var oisst = ee.ImageCollection('NOAA/CDR/OISST/V2_1');
var meanSST = oisst.select('sst')
  .filter(ee.Filter.date('2024-01-01', '2025-01-01'))
  .mean().multiply(0.01).clip(ocean);          // scale 0.01 -> °C
Map.addLayer(meanSST, {min: 24, max: 31, palette: sstPalette}, 'Mean SST 2024 (°C)');

// ===== 3. WARM-SEASON SST ANOMALY (bleaching risk snapshot) ========
var meanAnom = oisst.select('anom')
  .filter(ee.Filter.date('2024-01-01', '2024-04-01'))   // Pacific warm season
  .mean().multiply(0.01).clip(ocean);
Map.addLayer(meanAnom, {min: -2, max: 2, palette: anomPalette},
             'SST anomaly Jan-Mar 2024 (°C vs normal)');

// ===== 4. TIME SERIES: monthly SST anomaly (the bleaching signal) ==
var startMonth = ee.Date('2015-01-01');
var nMonths = ee.Date(Date.now()).difference(startMonth, 'month').floor();
var monthlyAnom = ee.ImageCollection.fromImages(
  ee.List.sequence(0, nMonths).map(function (m) {
    var t0 = startMonth.advance(m, 'month');
    return oisst.select('anom')
      .filter(ee.Filter.date(t0, t0.advance(1, 'month')))
      .mean().multiply(0.01).set('system:time_start', t0.millis());
  })
);
print(ui.Chart.image.series({
  imageCollection: monthlyAnom, region: ocean,
  reducer: ee.Reducer.mean(), scale: 25000
}).setOptions({
  title: 'Monthly sea surface temperature anomaly around ' + COUNTRY + ' (2015-present)',
  vAxis: {title: 'SST anomaly (°C vs normal)'}, hAxis: {title: 'Date'},
  lineWidth: 1.5, series: {0: {color: 'b2182b'}}, legend: {position: 'none'}
}));
print('Tip: anomalies sustained above +1°C for weeks indicate coral bleaching heat stress.');

// ===== 5. EXPORT (optional) ========================================
Export.image.toDrive({
  image: meanSST,
  description: COUNTRY.replace(/[ ,]/g, '_') + '_MeanSST_2024',
  folder: 'GEE_Workshop_2026',
  fileNamePrefix: COUNTRY.replace(/[ ,]/g, '_') + '_mean_sst_2024',
  scale: 25000, region: ocean.bounds(), maxPixels: 1e13
});
